sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala

```diff
@@ -704,6 +704,31 @@ class ExplainSuiteAE extends ExplainSuiteHelper with EnableAdaptiveExecutionSuit
         "Bucketed: false (bucket column(s) not read)")
     }
   }
+
+  test("SPARK-36795: Node IDs should not be duplicated when InMemoryRelation Present") {
+    withTempView("t1", "t2") {
+      Seq(1).toDF("k").write.saveAsTable("t1")
+      Seq(1).toDF("key").write.saveAsTable("t2")
+      spark.sql("SELECT * FROM t1").persist()
+      val query = "SELECT * FROM (SELECT * FROM t1) join t2 " +
+        "ON k = t2.key"
+      val df = sql(query).toDF()
+
+      df.collect()
+      checkKeywordsExistsInExplain(df, FormattedMode,
+        """   * BroadcastHashJoin Inner BuildLeft (12)
+          |   :- BroadcastQueryStage (8)
+          |   :  +- BroadcastExchange (7)
+          |   :     +- * Filter (6)
+          |   :        +- * ColumnarToRow (5)
```

Comment:
nit but can we just extract `ColumnarToRow (num)` and `InMemoryRelation (num)` by regex, and assert that the ids are different instead? Showing the explain here would more likely make other PRs change this test together.

Reply:
Ok. changed the test to regex that extracts the node ids and asserts they are different.